Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/NodeReferencing.php

```diff
@@ -112,52 +102,55 @@ private function handleSetSerializedNodeReferences(
             $sourceNodeAggregate,
             $command->sourceOriginDimensionSpacePoint
         );
-        $this->requireNodeTypeToDeclareReference($sourceNodeAggregate->nodeTypeName, $command->referenceName);
+
+        $sourceNodeType = $this->requireNodeType($sourceNodeAggregate->nodeTypeName);
+        $events = [];
 
         $this->requireNodeTypeToAllowNumberOfReferencesInReference(
             $command->references,
-            $command->referenceName,
             $sourceNodeAggregate->nodeTypeName
         );
 
-        foreach ($command->references as $reference) {
-            assert($reference instanceof SerializedNodeReference);
-            $destinationNodeAggregate = $this->requireProjectedNodeAggregate(
-                $contentGraph,
-                $reference->targetNodeAggregateId
-            );
-            $this->requireNodeAggregateToNotBeRoot($destinationNodeAggregate);
-            $this->requireNodeAggregateToCoverDimensionSpacePoint(
-                $destinationNodeAggregate,
-                $command->sourceOriginDimensionSpacePoint->toDimensionSpacePoint()
-            );
-            $this->requireNodeTypeToAllowNodesOfTypeInReference(
-                $sourceNodeAggregate->nodeTypeName,
-                $command->referenceName,
-                $destinationNodeAggregate->nodeTypeName
-            );
-        }
-
-        $sourceNodeType = $this->requireNodeType($sourceNodeAggregate->nodeTypeName);
-        $scopeDeclaration = $sourceNodeType->getReferences()[$command->referenceName->value]['scope'] ?? '';
-        $scope = PropertyScope::tryFrom($scopeDeclaration) ?: PropertyScope::SCOPE_NODE;
+        foreach ($command->references as $referencesByName) {
+            $referenceName = $referencesByName->referenceName;
+            $this->requireNodeTypeToDeclareReference($sourceNodeAggregate->nodeTypeName, $referenceName);
+            $scopeDeclaration = $sourceNodeType->getReferences()[$referenceName->value]['scope'] ?? '';
+            $scope = PropertyScope::tryFrom($scopeDeclaration) ?: PropertyScope::SCOPE_NODE;
+            // TODO: Optimize affected sets into one event
```

Comment:
Is there a specific reason why we _don't_?
The current state would be like publishing multiple `NodePropertiesWereSet` events for each property.

Couldn't we just build up the `$referencesByName` array and then build the event outside of the loop?

Reply:
Adjusted:)